### DISCUSSION TITLE
SSE-C feature ideation

### DISCUSSION THREAD

**Discussion by adasauce** [seaweedfs/seaweedfs#5361] [Ideas]
Hi Folks!

I have a stateless implementation of SSE-C working to encrypt/decrypt objects through the s3 api at rest using AES keys provided through the headers as described here: https://docs.aws.amazon.com/AmazonS3/latest/userguide/ServerSideEncryptionCustomerKeys.html

It prepends the aes initialization vector into the object's encrypted body on disk, and is randomly generated per object. So it takes the key supplied by the user for the request at `x-amz-server-side-encryption-customer-key` and sets up an io stream through a cypher io reader to decrypt it by taking a bite from the reader of the size of `aes.BlockSize`.

The goal of the implementation is that it doesn't have much of a surface impact on the whole implementation. No APIs need to change drastically to support it, just when the headers are present for SSE-C to function it pipes the readers and writers through relevant encryption/decryption streams that it knows how to build itself. 

There are a few cases with the `ceph/s3-tests` repo that are failing and I'm not sure the best way to deal with it. The biggest hurdle is that I think we'll have to store the key md5 in the filer metadata to detect objects that are keyed with encryption.

For example, a HEAD request to an object that was encrypted with a customer key is supposed to raise an error that the key doesn't match what the object was stored with:

```text
If you encrypt an object by using server-side encryption with customer-provided encryption keys (SSE-C) when you
 store the object in Amazon S3, then when you retrieve the metadata from the object, you must use the following 
headers to provide the encryption key for the server to be able to retrieve the object's metadata. The headers are:

    x-amz-server-side-encryption-customer-algorithm
    x-amz-server-side-encryption-customer-key
    x-amz-server-side-encryption-customer-key-MD5
```

So without reading the object's actual data the API  in some places expects to be able to know whether the user has supplied a valid key to access the object. 

Personally, I think if the object was encrypted, and it was requested without encryption, the encrypted body should act like the actual body of the object. The docs are very clear that when you use server side encryption with customer keys, it's the end user's responsibility to manage those keys and use them to access the data (and if they're lost, the data is effectively lost). From a security point of view I think it makes more sense to serve an encrypted blob than to say "you supplied the wrong key, try again".

I have a number of passing tests now, but they are all the tests around encrypting / downloading / decrypting objects and don't think much about querying bucket policies or use AWS kms features. There are also some bucket policy features that aren't implemented, and I haven't thought much about the best way.

---

I plan to open a PR at some point if this kind of feature can be added on as cleanly as it looks like it should, but before making too many assumptions I want to open the conversation about the overall goals and what kinds of things we do or don't want to be storing to enable this kind of feature.

On a quick look, there would be a lot of changes within the filer and pbs to support storing the key's md5 sum along with the objects in the existing attributes data structures, and I think the filer shouldn't really care about this feature.

**Comment by chrislusf** [comment]
This is a nice feature to add. Current filer gRPC APIs should be able to support the additional metadata required for this feature.

**Reply by adasauce** [reply]
I was looking around a bit more tonight. I think we can just leverage the existing SaveAmzMetaData to bring in the encryption metadata we care about (key md5 + encryption algorithm headers) without making any model changes. 

Thanks for the hint.

**Comment by kmlebedev** [comment]
Hi,
It seems to me that the implementation of external KMS https://github.com/nsmithuk/local-kms  and AES GCM https://github.com/rclone/rclone/pull/7084 is simpler and more manageable and does not require much modification from S3 and can work with any storage

**Reply by adasauce** [reply]
yeah, I'm currently not looking into any kind of kms or default sse-s3 encryption, only adding support for SSE-C customer supplied keys. these both seem like interesting approaches though.  

It would be prudent look into having configurable kms backends to get and fetch keys to not add a bunch of complexity within the project and risk fumbling the implementation of secret storage to have some security issues down the road.

**Comment by kvaps** [comment]
Hey @adasauce how is it going?

Do you have any code highlights for this feature implementation? Do you need help with it?

**Reply by Thirimash** [reply]
Yea imo its one of the most important feature actualy my company move from seaweedfs to another tehcnology due to lack of KMS.

**Reply by adasauce** [reply]
I made it a good way into this when the implementation before the project that I needed it for stopped. I will require it in the future again, but it's stalled for now for me.

The stickiest part was that I was trying to transparently encrypt/decrypt the data from volumes at the api but there was an issue with content verification because the hashed value returned from the encrypted bit didn't match obviously. I was going down the path of a simple encoding format to embed the original hash as a header within the stored data so that it could be verified when I stopped working on it. I had also identified that file size calculation was also going to be an issue to tackle in terms of both reporting to the user and internal bean counting.

Things may have also evolved architecturally in seaweed since I was seriously looking at it so I was planning on taking my initial notes and going from scratch again rather than trying to rebase on anything new.

**Reply by chrislusf** [reply]
added https://github.com/seaweedfs/seaweedfs/pull/7143

@adasauce maybe you want to test it before I merge it?

**Reply by Thirimash** [reply]
hello @chrislusf when you will implement it and make documentation to use it ? im asking this question to mabe do not move to another solution

**Reply by chrislusf** [reply]
Should be pretty soon after this https://github.com/seaweedfs/seaweedfs/pull/7144 for SSE-KMS.

But this would need someone to test it.

**Reply by chrislusf** [reply]
Update: SSE-C is merged. Welcome to test!

SSE-KMS is coming also. Any specific KMS provider to support first? As long as someone can test, it can be prioritized.

**Reply by Thirimash** [reply]
hello imho preffered option will be support open source projects at first. So maby try to test it at Open Bao its fork from hashicorp vault before they changed license and its one of the most popular and easy  KMS system . Do you have any documentation or instruction how to set up connection ?

**Reply by chrislusf** [reply]
Here are wiki pages:
- https://github.com/seaweedfs/seaweedfs/wiki/Server-Side-Encryption
- http://github.com/seaweedfs/seaweedfs/wiki/Server-Side-Encryption-SSE-KMS
- https://github.com/seaweedfs/seaweedfs/wiki/Server-Side-Encryption-SSE-C

**Reply by Thirimash** [reply]
Thanks for it @chrislusf , when you planing to implement KMS in official realease ?

**Reply by Thirimash** [reply]
@chrislusf any plans for implemeting it in new official patch ?

**Reply by chrislusf** [reply]
this is merged today and should be released tomorrow.
